src/useImageDimensions.ts

```diff
@@ -0,0 +1,59 @@
+import {useEffect, useState} from 'react'
+import {Image, ImageRequireSource} from 'react-native'
+
+export interface URISource {
+  uri: string
+}
+
+export type Source = ImageRequireSource | URISource
+
+export interface ImageDimensions {
+  dimensions?: {width: number; height: number; aspectRatio: number}
+  error?: Error
+  loading: boolean
+}
+
+/**
+ * @param source either a remote URL or a local file resource.
+ * @returns original image dimensions (width, height and aspect ratio).
+ */
+export function useImageDimensions(source: Source): ImageDimensions {
+  const [result, setResult] = useState<ImageDimensions>({loading: true})
+
+  useEffect(() => {
+    try {
+      if (typeof source === 'number') {
+        const {width, height} = Image.resolveAssetSource(source)
+
+        setResult({
+          dimensions: {width, height, aspectRatio: width / height},
+          loading: false,
+        })
+
+        return
+      }
+
+      if (typeof source === 'object' && source.uri) {
+        setResult({loading: true})
+
+        Image.getSize(
+          source.uri,
+          (width, height) =>
+            setResult({
+              dimensions: {width, height, aspectRatio: width / height},
+              loading: false,
+            }),
+          (error) => setResult({error, loading: false}),
+        )
+
+        return
+      }
+
+      throw new Error('not implemented')
+    } catch (error) {
+      setResult({error, loading: false})
+    }
+  }, [source])
```

Comment:
A question here: if the `source` is an object, wouldn't it be changed every render, triggering the warning: `Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or **one of the dependencies changes on every render.**`? This is because `useEffect` hook does shallow comparison, not deep comparison.

Reply:
It depends on how you pass in the `source`, as long as the same object is being passed each render cycle it will not trigger the `useEffect` to re-run. The one case I can see would be problematic is if people pass the url as such:

```
useImageDimensions({ uri: 'https://...' })
```

We could help with this by doing something like:

```diff
-  }, [source])
+  }, [source?.uri ?? source])
```

Reply:
@LinusU Yes. That would definitely do the trick. 👍 👍 👍

Reply:
Also, it would be nice to update the doc too. This would be a common trap that perplex users if they don't know where to look for e.g. when they are working in a complex component with lots of things to already worry them, it'll take effort to narrow down where an infinite loop might occur.